File: LeetCode/37_解数独.java
import java.util.*;

/*
    37. 解数独
    编写一个程序，通过已填充的空格来解决数独问题。

    编写一个程序，通过已填充的空格来解决数独问题。

    执行用时：7 ms, 在所有 Java 提交中击败了52.89%的用户
    内存消耗：37.4 MB, 在所有 Java 提交中击败了20.59%的用户
 */


class Solution {
    boolean finish = false;

    public void solveSudoku(char[][] board) {
        solveSudoku(board, 0, 0);
    }

    public void solveSudoku(char[][] board, int x, int y) {
        if (board[x][y] == '.') {
            for (char i = '1'; i <= '9'; i++) {
                if (check(board, x, y, i)) {
                    board[x][y] = i;
                    if (x == 8 && y == 8) finish = true;
                    if (y < 8) solveSudoku(board, x, y + 1);
                    else if (x < 8) solveSudoku(board, x + 1, 0);
                    if (finish) break;
                    board[x][y] = '.';
                }
            }
        } else {
            if (x == 8 && y == 8) {
                finish = true;
                return;
            }
            if (y < 8) solveSudoku(board, x, y + 1);
            else if (x < 8) solveSudoku(board, x + 1, 0);
        }
    }

    public boolean check(char[][] board, int x, int y, char value) {
        for (int i = 0; i < 9; i++) {
            if (board[i][y] == value) return false;
            if (board[x][i] == value) return false;
        }

        int groupX = x / 3, groupY = y / 3;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (board[groupX * 3 + i][groupY * 3 + j] == value) return false;
            }
        }

        return true;
    }
}


class LeetCode {
    public static void main(String[] args) {
        Solution s = new Solution();
        char[][] board = new char[][]{
                {'5', '3', '.', '.', '7', '.', '.', '.', '.'},
                {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
                {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
                {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
                {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
                {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
                {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
                {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
                {'.', '.', '.', '.', '8', '.', '.', '7', '9'}
        };
        s.solveSudoku(board);
        System.out.println(Arrays.deepToString(board));
    }
}

File: LeetCode/37_解数独.py
# -*- coding: UTF-8 -*-
"""=================================================
Problem: 37. 解数独
Website: https://leetcode.cn/problems/sudoku-solver/description/
Difficulty: 困难
Author: ZYZ
Language: Python3
Result: Accepted
Runtime: 304 ms
Memory Usage: 16.1 MB
=================================================="""
from typing import List, Dict, Set
from collections import defaultdict, deque, Counter
from itertools import product
from pprint import pprint
import copy

"""
编写一个程序，通过填充空格来解决数独问题。

数独的解法需 遵循如下规则：

数字 1-9 在每一行只能出现一次。
数字 1-9 在每一列只能出现一次。
数字 1-9 在每一个以粗实线分隔的 3x3 宫内只能出现一次。
数独部分空格内已填入了数字，空白格用 '.' 表示。
"""


class Solution:
    solve = False

    def solveSudoku(self, board: List[List[str]]) -> None:
        self.dfs(board)

    def dfs(self, board: List[List[str]]):
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    if i == 8 and j == 8:
                        self.solve = True
                    continue
                suitable = self.candidate(board, i, j)
                for suit in suitable:
                    board[i][j] = suit
                    self.dfs(board)
                    if self.solve:
                        return
                    board[i][j] = '.'
                return

    def candidate(self, board: List[List[str]], i: int, j: int) -> List[str]:
        row = set(board[i])
        column = set([x[j] for x in board])
        block = set()
        for x in range(i // 3 * 3, i // 3 * 3 + 3):
            for y in range(j // 3 * 3, j // 3 * 3 + 3):
                block.add(board[x][y])
        fill = (row | column | block)
        return list(set([str(z) for z in range(1, 10)]) - fill)


if __name__ == '__main__':
    solution = Solution()
    answer = [["5", "3", "4", "6", "7", "8", "9", "1", "2"],
              ["6", "7", "2", "1", "9", "5", "3", "4", "8"],
              ["1", "9", "8", "3", "4", "2", "5", "6", "7"],
              ["8", "5", "9", "7", "6", "1", "4", "2", "3"],
              ["4", "2", "6", "8", "5", "3", "7", "9", "1"],
              ["7", "1", "3", "9", "2", "4", "8", "5", "6"],
              ["9", "6", "1", "5", "3", "7", "2", "8", "4"],
              ["2", "8", "7", "4", "1", "9", "6", "3", "5"],
              ["3", "4", "5", "2", "8", "6", "1", "7", "9"]]
    question = [["5", "3", ".", ".", "7", ".", ".", ".", "."],
                ["6", ".", ".", "1", "9", "5", ".", ".", "."],
                [".", "9", "8", ".", ".", ".", ".", "6", "."],
                ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
                ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
                ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
                [".", "6", ".", ".", ".", ".", "2", "8", "."],
                [".", ".", ".", "4", "1", "9", ".", ".", "5"],
                [".", ".", ".", ".", "8", ".", ".", "7", "9"]]
    print(solution.solveSudoku(question))
    pprint(question)
